### PR TITLE
Introduce metadata for all add-ons

### LINKED ISSUE #2058: Introduce add-on metadata
This is a follow-up on https://github.com/eclipse-archived/smarthome/issues/4503 suggesting to define a `OH-INF/addon.xml` file that could include different metadata about the add-on.

In the case of bindings, this should be a replacement for the `binding.xml` file.

Content of the `OH-INF/addon.xml` file should be:

- `name` - mandatory - The add-on name, no reference to specific solution (e.g. openHAB). Unclear if type should be included ("Air Quality" vs. "Air Quality Binding")
- `description` - mandatory - A description of the purpose, specialization and limitations of the add-on, should be between 160-240 characters.
- `type` - mandatory - One out of a counted list, e.g. "Binding", "Automation", ...
- `keywords` - optional but recommended - Keywords (categories) out of an open list to filter against, e.g. "heating, lighting, graphing"
- `connection` - optional - highlight if a binding/automation/... interacts locally or if it connects to an online service (seldom or continuously)
- `countries` - optional - a comma-separated list of ISO 3166 2-letter country codes for which this add-on is applicable

The metadata must be made available through the REST API, so that it can be consumed by the UI.

### PR DESCRIPTION
**Pull request by J-N-K**
Closes #2058
Depends on #2994

- [x] Adjust add-ons in openhab-addons (https://github.com/openhab/openhab-addons/pull/13289)
- [x] Adjust zwave binding (https://github.com/openhab/org.openhab.binding.zwave/pull/1798, https://github.com/openhab/org.openhab.binding.zwave/pull/1839)
- [x] Adjust zigbee binding (https://github.com/openhab/org.openhab.binding.zigbee/pull/793)
- [x] Adjust add-ons on openhab-webui (https://github.com/openhab/openhab-webui/pull/1468)
- [x] Adjust Developer documentation (https://github.com/openhab/openhab-docs/pull/1872)
- [x] Adjust add-on template

This removes the `binding.xml` and replaces it with an `addon.xml` which allows to provide metadata for all add-ons. Also contains necessary changes in other components.

### CODE REVIEW

**Comment by lsiepel** [comment]
I don’t see a PR for updating the developer documentation / binding  template etc. Am I overlooking something? I hope so.

**Comment by wborn** [comment]
The i18n-plugin also needs to be updated as it is reads Binding XML info for generating translations. It will probably still compile locally because its bundle [dependency](https://github.com/openhab/openhab-core/blob/8f745b0d3d5d263cffcd50b2a98989486af5415f/tools/i18n-plugin/pom.xml#L52-L56) still exists in your local and the OH snapshot repo.

**Comment by J-N-K** [comment]
The i18n-maven-plugin changes are already part of this PR.

**Comment by jlaur** [comment]
What is the next step for this PR? Perhaps now would be a good time to get it finished so we have a full release cycle for testing, assuming it would make it into the first 4.0 milestone.

Before asking @J-N-K to resolve merge conflicts (also in related PR's) and preparing a PR for the Zigbee binding, perhaps a @openhab/core-maintainers could have a look and write some kind of statement about "conditional approval", so that Jan won't have to resolve conflicts and rebase too many times? @kaikreuzer - it seems this was suggested by you in #2058, so perhaps you could have a look and comment? 🙂

**Comment by kaikreuzer** [comment]
Good point @jlaur! Yes, I'll have a look at it the next days (latest by the end of the week). Thanks @J-N-K for rebasing it already.

**Comment by kaikreuzer** [comment]
> Adjust zigbee binding

I assume it'll be pretty quick and straight forward to address this as well as soon as this PR here is merged, right?

**Comment by J-N-K** [comment]
Yes, a second zwave and zigbee PR are easy, as I said in webui: for openhab-addons I need a short time.

**Comment by kaikreuzer** [comment]
Could you please fix the project name in the 2 moved `.project` files? (I know, you are not using Eclipse, but for those who do this is important. 😄 )

**Comment by J-N-K** [comment]
Done

File: bundles/org.openhab.core.addon.xml/src/main/java/org/openhab/core/addon/xml/internal/AddonInfoReader.java
Comment (kaikreuzer) [inline bundles/org.openhab.core.addon.xml/src/main/java/org/openhab/core/addon/xml/internal/AddonInfoReader.java:75]: What has happened to the `service-id`?

File: bundles/org.openhab.core.addon.xml/src/main/java/org/openhab/core/addon/xml/internal/AddonInfoReader.java
Comment (kaikreuzer) [inline bundles/org.openhab.core.addon.xml/src/main/java/org/openhab/core/addon/xml/internal/AddonInfoReader.java:85]: What is the `config-id`?

File: bundles/org.openhab.core.addon.xml/src/main/java/org/openhab/core/addon/xml/internal/AddonInfoConverter.java
Comment (kaikreuzer) [inline bundles/org.openhab.core.addon.xml/src/main/java/org/openhab/core/addon/xml/internal/AddonInfoConverter.java:71]:
```suggestion
        String id = requireNonEmpty(attributes.get("id"), "Add-on id attribute is null or empty");
```

File: bundles/org.openhab.core.test.magic/src/main/resources/OH-INF/addon/addon.xml
Comment (kaikreuzer) [inline bundles/org.openhab.core.test.magic/src/main/resources/OH-INF/addon/addon.xml:6]: `core` doesn't seem to be a valid value - shouldn't this be `binding` here?

File: bundles/org.openhab.core/src/main/java/org/openhab/core/addon/AddonI18nLocalizationService.java
Comment (kaikreuzer) [inline bundles/org.openhab.core/src/main/java/org/openhab/core/addon/AddonI18nLocalizationService.java:47]:
```suggestion
     * @param addonInfo the add-on info that should be localized
```

File: bundles/org.openhab.core/src/main/java/org/openhab/core/addon/AddonI18nLocalizationService.java
Comment (kaikreuzer) [inline bundles/org.openhab.core/src/main/java/org/openhab/core/addon/AddonI18nLocalizationService.java:44]:
```suggestion
     * Localizes an add-on info.
```

File: bundles/org.openhab.core/src/main/java/org/openhab/core/addon/AddonI18nLocalizationService.java
Comment (kaikreuzer) [inline bundles/org.openhab.core/src/main/java/org/openhab/core/addon/AddonI18nLocalizationService.java:49]:
```suggestion
     * @return a localized add-on info on success, a non-localized one on error (e.g. no translation is found).
```

**Review by kaikreuzer — CHANGES_REQUESTED** [review]
A few first (very small) findings - more to come tomorrow!

File: bundles/org.openhab.core.addon.xml/src/main/java/org/openhab/core/addon/xml/internal/AddonInfoReader.java
Comment (J-N-K) [reply bundles/org.openhab.core.addon.xml/src/main/java/org/openhab/core/addon/xml/internal/AddonInfoReader.java:75]: Originally I planned to change it to `config-id` and later decided against that. I have reverted this changed (same answer below).

File: bundles/org.openhab.core/src/main/java/org/openhab/core/addon/AddonInfo.java
Comment (kaikreuzer) [inline bundles/org.openhab.core/src/main/java/org/openhab/core/addon/AddonInfo.java:98]:
```suggestion
     * Returns a human-readable name for the add-on (e.g. "HUE Binding").
```

File: bundles/org.openhab.core/src/main/java/org/openhab/core/addon/AddonInfo.java
Comment (kaikreuzer) [inline bundles/org.openhab.core/src/main/java/org/openhab/core/addon/AddonInfo.java:100]:
```suggestion
     * @return a human-readable name for the add-on (neither null, nor empty)
```

File: bundles/org.openhab.core/src/main/java/org/openhab/core/addon/AddonInfo.java
Comment (kaikreuzer) [inline bundles/org.openhab.core/src/main/java/org/openhab/core/addon/AddonInfo.java:115]:
```suggestion
     * Returns a human-readable description for the add-on
```

File: bundles/org.openhab.core/src/main/java/org/openhab/core/addon/AddonInfo.java
Comment (kaikreuzer) [inline bundles/org.openhab.core/src/main/java/org/openhab/core/addon/AddonInfo.java:118]:
```suggestion
     * @return a human-readable description for the add-on
```

File: bundles/org.openhab.core/src/main/java/org/openhab/core/addon/AddonInfo.java
Comment (kaikreuzer) [inline bundles/org.openhab.core/src/main/java/org/openhab/core/addon/AddonInfo.java:125]:
```suggestion
     * Returns the author of the add-on (e.g. "openHAB").
```

File: bundles/org.openhab.core/src/main/java/org/openhab/core/addon/AddonInfo.java
Comment (kaikreuzer) [inline bundles/org.openhab.core/src/main/java/org/openhab/core/addon/AddonInfo.java:127]:
```suggestion
     * @return the author of the add-on (could be null or empty)
```

**Review by kaikreuzer — CHANGES_REQUESTED** [review]
Many thanks @J-N-K!
Looks all perfect to me - all I could find are afew remaining missed replacements in the Javadoc.

**Comment by J-N-K** [comment]
Thanks, I also prepared the missing PR for zwave and zigged (which are quite straight-forward) and updated the PR in openhab-addons (which is also mostly copying and addition of the new add-ons like persistence). What is missing is a proper review of the webui-PR regarding the changes in MainUI.

**Review by kaikreuzer — APPROVED** [review]
> What is missing is a proper review of the webui-PR regarding the changes in MainUI.

I can't really review the code, but I will test if it still works in general. If there are small things to correct, I guess @ghys can also comment/fix it later on.

So would you say that from your pov everything should then be ready to be merged today already?

**Comment by J-N-K** [comment]
I think it's fine now, I removed some leftovers in addons, otherwise I would say: Go! If something comes up I can fix it on short notice.

**Comment by kaikreuzer** [comment]
Excellent, then let us go for it!

**Comment by J-N-K** [comment]
@kaikreuzer Can you upload the XSD so it is available for validation? I'll prepare a PR to change the static analysis.

**Comment by kaikreuzer** [comment]
It is all in git. :-) 
XSD uploaded here: https://github.com/openhab/website/pull/387

File: bundles/org.openhab.core.addon.xml/addon-1.0.0.xsd
Comment (jlaur) [inline bundles/org.openhab.core.addon.xml/addon-1.0.0.xsd:19]: Sorry for commenting code already merged, but thought it would be easiest for proper context. Should it be mentioned that it's not needed to use "openHAB" as author here? At least it seems none of the add-ons in the openhab-addons repository have this, and the skeleton script doesn't create it.

Comment (kaikreuzer) [reply bundles/org.openhab.core.addon.xml/addon-1.0.0.xsd:19]: Maybe it would be the right time to drop this field now altogether - we more or less only kept it for backward-compatibility, but as we are breaking this now anyhow, I think we could remove it from the XSD.

Comment (ghys) [reply bundles/org.openhab.core.addon.xml/addon-1.0.0.xsd:19]: The marketplace add-on service will build the add-on with what Discourse returns as the author of the topic as the add-on author:

https://github.com/openhab/openhab-core/blob/8d64ecfd8db99844d79a776fef37cbe527c1f0c6/bundles/org.openhab.core.addon.marketplace/src/main/java/org/openhab/core/addon/marketplace/internal/community/CommunityMarketplaceAddonService.java#L334-L336

and the Karaf add-on service will put the constant "openHAB" in that field (and set it as "verified"):

https://github.com/openhab/openhab-core/blob/8f9dafd1d68a92248cd39414145175952d61cd46/bundles/org.openhab.core.karaf/src/main/java/org/openhab/core/karaf/internal/KarafAddonService.java#L133-L134

The author and the checkmark are displayed in the UI, notably in the information tables in each add-on page. For "verified" add-ons it's also displayed in the list.

Comment (kaikreuzer) [reply bundles/org.openhab.core.addon.xml/addon-1.0.0.xsd:19]: Thanks @ghys, this means that this field is definitely deprecated then.
@J-N-K Would you agree that this is a good moment to remove it from the XSD then?

Comment (J-N-K) [reply bundles/org.openhab.core.addon.xml/addon-1.0.0.xsd:19]: Yes

**Pull request merged**
